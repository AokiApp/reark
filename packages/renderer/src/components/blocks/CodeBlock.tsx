import { BlockInnerComponent } from "../../types";
import { CODE_LANGUAGE } from "../../constants/codeLanguage";
import { Text } from "./Text";

export const CodeBlock: BlockInnerComponent = ({ block }) => {
  const code = block.code;
  if (!code) {
    return null;
  }
  const codeLanguage = code?.style?.language
    ? CODE_LANGUAGE[code.style.language]
    : "Unknown";

  return (
    <div className="reark-codeblock">
      <div className="reark-codeblock__language">{codeLanguage}</div>
      <pre className="reark-codeblock__pre">
        <Text elements={code.elements} style={code.style} />
      </pre>
    </div>
  );
};
// todo : copy button
