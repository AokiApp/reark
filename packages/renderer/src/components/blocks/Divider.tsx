import { BlockInnerComponent } from "../../types";

export const Divider: BlockInnerComponent = () => {
  return <hr className="reark-divider" />;
};
