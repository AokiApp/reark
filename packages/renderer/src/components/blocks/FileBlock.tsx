import { BlockInnerComponent } from "../../types";

/**
 * FileBlock: Renders a file block (block_type: 23).
 * - Shows file name and download link if available.
 * - This block is usually paired with a ViewBlock (block_type: 33) for visual representation.
 * - If rendered directly, provides basic file info.
 */
export const FileBlock: BlockInnerComponent = ({ block }) => {
  const file = block.file;
  if (!file) {
    return (
      <div className="reark-file-block reark-file-block--empty">
        No file data
      </div>
    );
  }

  return (
    <div className="reark-file-block">
      <div>
        <strong>File:</strong> {file.name || "(no name)"}
      </div>
      {file.token ? (
        <a
          href={`https://open.feishu.cn/drive/file/${file.token}`}
          target="_blank"
          rel="noopener noreferrer"
          className="reark-file-block__file-link"
        >
          Download / Open
        </a>
      ) : (
        <span className="reark-file-block__file-link--disabled">
          (No file token)
        </span>
      )}
    </div>
  );
};
