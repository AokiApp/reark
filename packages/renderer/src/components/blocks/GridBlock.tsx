import { BlockInnerComponent } from "../../types";
import { BlockComponent } from "../BlockComponent";

/**
 * GridBlock: Renders a grid container with a specified number of columns.
 * Each child is expected to be a GridColumnBlock.
 */
export const GridBlock: BlockInnerComponent = ({ block }) => {
  const grid = block.grid;
  if (!grid || !block.children || !Array.isArray(block.children)) {
    return null;
  }

  // column_size: number of columns (2-4)
  //const columnSize = grid.column_size || 2;

  // Render children by block reference (blockId), allowing for block reference resolution.
  return (
    <div className="reark-grid-wrapper">
      <div
        className="reark-grid"
        style={{
          display: "flex",
          //gridTemplateColumns: `repeat(${columnSize}, 1fr)`,
          gap: "16px",
        }}
      >
        {block.children.map((childId) => (
          // BlockComponent resolves and renders the block by its ID (block reference)
          <BlockComponent key={childId} blockId={childId} />
        ))}
      </div>
    </div>
  );
};
