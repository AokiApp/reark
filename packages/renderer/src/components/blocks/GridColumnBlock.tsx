import { BlockInnerComponent } from "../../types";
import { BlockComponent } from "../BlockComponent";

/**
 * GridColumnBlock: Renders a single column in a grid, with optional width ratio.
 * Children can be any block type.
 */
export const GridColumnBlock: BlockInnerComponent = ({ block }) => {
  const gridColumn = block.grid_column;
  if (!block.children || !Array.isArray(block.children)) {
    return null;
  }

  // width_ratio: ratio of this column to the whole grid (1-99)
  const widthRatio = gridColumn?.width_ratio || 1;

  // Render children by block reference (blockId), allowing for block reference resolution.
  return (
    <div
      className="reark-grid-column"
      style={{
        flex: widthRatio,
        minWidth: 0,
        display: "flex",
        flexDirection: "column",
        gap: "8px",
      }}
    >
      {block.children.map((childId) => (
        // BlockComponent resolves and renders the block by its ID (block reference)
        <BlockComponent key={childId} blockId={childId} />
      ))}
    </div>
  );
};
