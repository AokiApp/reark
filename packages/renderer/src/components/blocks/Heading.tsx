import { BlockInnerComponent } from "../../types";
import { Text } from "./Text";

type HeadingLevel = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9;

const HeadingBase: BlockInnerComponent<{
  level: HeadingLevel;
}> = ({ block, level }) => {
  const headingKey = `heading${level}` as const;
  const headingData = block[headingKey];

  if (!headingData?.elements) {
    return null;
  }

  return (
    <div
      className={`reark-heading reark-heading--${level}`}
      id={block.block_id}
    >
      <Text elements={headingData.elements} style={headingData.style} />
    </div>
  );
};

export const Heading1: BlockInnerComponent = ({ block }) => (
  <HeadingBase block={block} level={1} />
);
export const Heading2: BlockInnerComponent = ({ block }) => (
  <HeadingBase block={block} level={2} />
);
export const Heading3: BlockInnerComponent = ({ block }) => (
  <HeadingBase block={block} level={3} />
);
export const Heading4: BlockInnerComponent = ({ block }) => (
  <HeadingBase block={block} level={4} />
);
export const Heading5: BlockInnerComponent = ({ block }) => (
  <HeadingBase block={block} level={5} />
);
export const Heading6: BlockInnerComponent = ({ block }) => (
  <HeadingBase block={block} level={6} />
);
export const Heading7: BlockInnerComponent = ({ block }) => (
  <HeadingBase block={block} level={7} />
);
export const Heading8: BlockInnerComponent = ({ block }) => (
  <HeadingBase block={block} level={8} />
);
export const Heading9: BlockInnerComponent = ({ block }) => (
  <HeadingBase block={block} level={9} />
);
