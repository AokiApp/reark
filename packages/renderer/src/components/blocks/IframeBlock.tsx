import React from "react";
import { BlockInnerComponent } from "../../types";

function decodeUrl(url: string): string {
  try {
    return decodeURIComponent(url);
  } catch {
    return url;
  }
}

export const IframeBlock: BlockInnerComponent = ({ block }) => {
  const component = block.iframe?.component;
  if (!component) return null;

  const { iframe_type, url } = component;
  const decodedUrl = decodeUrl(url);

  if (iframe_type === 15) {
    // YouTube埋め込み
    // URL例: https://www.youtube.com/watch?v=xxxx
    // 埋め込み用URLに変換
    let embedUrl = "";
    const match = decodedUrl.match(
      /(?:youtube\.com\/watch\?v=|youtu\.be\/)([A-Za-z0-9_-]+)/,
    );
    if (match && match[1]) {
      embedUrl = `https://www.youtube.com/embed/${match[1]}`;
    } else {
      embedUrl = decodedUrl;
    }
    return (
      <div className="reark-iframe">
        <iframe
          src={embedUrl}
          allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture"
          allowFullScreen
          title="YouTube Video"
        />
      </div>
    );
  } else {
    // 汎用iframe
    return (
      <div className="reark-iframe">
        <iframe
          src={decodedUrl}
          allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture"
          allowFullScreen
          title="Embedded Content"
        />
      </div>
    );
  }
};
