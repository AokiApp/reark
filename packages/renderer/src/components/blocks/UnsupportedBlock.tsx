import React from "react";

export interface UnsupportedBlockProps {
  type: number;
}

export const UnsupportedBlock: React.FC<UnsupportedBlockProps> = ({ type }) => (
  <div className="reark-unsupported-block">
    <strong>Unsupported block type: {type}</strong>
  </div>
);
