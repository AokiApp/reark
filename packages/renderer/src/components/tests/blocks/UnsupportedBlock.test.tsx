import { renderWithVRT } from "../test-utils/renderWithVRT";
import { UnsupportedBlock } from "../../blocks/UnsupportedBlock";

describe("UnsupportedBlock", () => {
  it("renders unsupported block message", async () => {
    const { container, vrt } = renderWithVRT(
      <UnsupportedBlock block={{ block_id: "dummy", block_type: 42 }} />,
    );
    expect(container.textContent).toMatch(/Unsupported block type: 42/);
    await vrt();
    // スナップショットテスト
    expect(container).toMatchSnapshot();
  });
});

it("renders unsupported block message with undefined block_type", async () => {
  const { container, vrt } = renderWithVRT(
    <UnsupportedBlock
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      block={{ block_id: "dummy", block_type: undefined as any }}
    />,
  );
  expect(container.textContent).toMatch(/Unsupported block type/);
  await vrt();
  expect(container).toMatchSnapshot();
});

it("renders unsupported block message with undefined block_id and block_type", async () => {
  const { container, vrt } = renderWithVRT(
    <UnsupportedBlock
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      block={{ block_id: undefined as any, block_type: undefined as any }}
    />,
  );
  expect(container.textContent).toMatch(/Unsupported block type/);
  await vrt();
  expect(container).toMatchSnapshot();
});
