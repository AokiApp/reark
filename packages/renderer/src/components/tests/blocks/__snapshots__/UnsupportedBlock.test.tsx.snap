// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UnsupportedBlock renders unsupported block message 1`] = `
<div>
  <div
    class="reark-unsupported-block"
  >
    <strong>
      Unsupported block type: 
      42
    </strong>
  </div>
</div>
`;

exports[`UnsupportedBlock renders unsupported block message 2`] = `
<div>
  <div
    class="reark-unsupported-block"
  >
    <strong>
      Unsupported block type: 
      42
    </strong>
  </div>
</div>
`;

exports[`renders unsupported block message with undefined type 1`] = `
<div>
  <div
    class="reark-unsupported-block"
  >
    <strong>
      Unsupported block type: 
    </strong>
  </div>
</div>
`;

exports[`renders unsupported block message with undefined type 2`] = `
<div>
  <div
    class="reark-unsupported-block"
  >
    <strong>
      Unsupported block type: 
    </strong>
  </div>
</div>
`;
